### PR TITLE
update for object lay out system with samples

### PR DESCRIPTION
**Pull request by liamcindy**

### CODE REVIEW

File: cocos/ui/UILayoutComponent.h [hunk 1]
```diff
@@ -29,68 +29,140 @@ THE SOFTWARE.
...
+        virtual float getRightMargin();
+        virtual void setRightMargin(float margin);
+
+        virtual float getTopMargin();
+        virtual void setTopMargin(float margin);
+
+        virtual float getButtomMargin();
+        virtual void setButtomMargin(float margin);
+
+#pragma endregion
```
Comment (minggo) [inline cocos/ui/UILayoutComponent.h]: It is not compatible in all compiles, and it will cause warnings in unsupported compilers.

Comment (liamcindy) [inline cocos/ui/UILayoutComponent.h]: Ok, i will remove that.

File: cocos/ui/UILayoutComponent.cpp [hunk 2]
```diff
@@ -54,253 +84,530 @@ namespace ui {
...
+    void LayoutComponent::refreshHorizontalMargin()
+    {
+        Node* parent = this->getOwnerParent();
+        if (parent == nullptr)
+            return;
+
+        Point ownerPoint = _owner->getPosition();
+        Point ownerAnchor = _owner->getAnchorPoint();
+        Size ownerSize = _owner->getContentSize();
+        Size parentSize = parent->getContentSize();
```
Comment (minggo) [inline cocos/ui/UILayoutComponent.cpp]: Please use reference instead, for example
Point &ownerPoint = _owner->getPosition();

File: cocos/ui/UILayoutComponent.h [hunk 3]
```diff
@@ -29,68 +29,140 @@ THE SOFTWARE.
...
+        virtual void setSizeHeight(float height);
+
+        virtual float getPercentHeight();
+        virtual void setPercentHeight(float percentHeight);
+
+        virtual bool isUsingStretchWidth();
+        virtual void setStretchWidthEnabled(bool isUsed);
+
+        virtual bool isUsingStretchHeight();
+        virtual void setStretchHeightEnabled(bool isUsed);
```
Comment (minggo) [inline cocos/ui/UILayoutComponent.h:128]: Can add const for getXXX functions?

File: cocos/ui/UILayoutComponent.h [hunk 4]
```diff
@@ -29,68 +29,140 @@ THE SOFTWARE.
...
+        virtual void setButtomMargin(float margin);
+
+#pragma endregion
+
+#pragma region Size & Percent
+        virtual Size getSize();
+        virtual void setSize(Size _size);
+
+        virtual bool isUsingPercentWidth();
+        virtual void setPercentWidthEnabled(bool isUsed);
```
Comment (minggo) [inline cocos/ui/UILayoutComponent.h:107]: These two functions are mismatch.

File: cocos/ui/UILayoutComponent.cpp (hunk 2 above)
Comment (liamcindy) [inline cocos/ui/UILayoutComponent.cpp]: OK

File: cocos/ui/UILayoutComponent.h (hunk 3 above)
Comment (liamcindy) [inline cocos/ui/UILayoutComponent.h:128]: OK

File: cocos/ui/UILayoutComponent.h (hunk 4 above)
Comment (liamcindy) [inline cocos/ui/UILayoutComponent.h:107]: virtual bool isUsingPercentWidth(); 
virtual void setPercentWidthEnabled(bool isUsed); 

change to set , get ?

File: cocos/ui/UILayoutComponent.h [hunk 5]
```diff
@@ -29,68 +29,140 @@ THE SOFTWARE.
...
+
+        virtual float getSizeWidth();
+        virtual void setSizeWidth(float width);
+
+        virtual float getPercentWidth();
+        virtual void setPercentWidth(float percentWidth);
```
Comment (minggo) [inline cocos/ui/UILayoutComponent.h:113]: What's the meaning of percent width?

Comment (liamcindy) [inline cocos/ui/UILayoutComponent.h:113]: Is that width percent better?

File: cocos/ui/UILayoutComponent.h (hunk 4 above)
Comment (liamcindy) [inline cocos/ui/UILayoutComponent.h:107]: @minggo 
virtual bool isUsingPercentWidth(); 
 virtual void setPercentWidthEnabled(bool isUsed); 

change to set , get ?

**Pull request closed without merging**